lib/mix/test/mix/tasks/test_test.exs

```diff
@@ -257,6 +258,75 @@ defmodule Mix.Tasks.TestTest do
     end)
   end
 
+  describe "--max-failures" do
+    test "validates flags" do
+      in_tmp("flag", fn ->
+        Mix.Tasks.New.run(["flag"])
+
+        File.cd!("flag", fn ->
+          flag_error_msg = "Expected type positive integer or \"infinity\", got "
+
+          assert mix(~w[test --max-failures 0]) =~ flag_error_msg
+          assert mix(~w[test --max-failures -1]) =~ flag_error_msg
+          assert mix(~w[test --max-failures 1.23]) =~ flag_error_msg
+
+          refute mix(~w[test --max-failures 100]) =~ flag_error_msg
+          refute mix(~w[test --max-failures infinity]) =~ flag_error_msg
+          refute mix(~w[test --max-failures :infinity]) =~ flag_error_msg
+        end)
+      end)
+    end
+
+    test "stops on max-failures - sequentially" do
+      in_fixture("max_failures", fn ->
+        assert mix(:sequential, ~w[test --max-failures 2]) =~ "\n3 tests, 2 failures\n"
+      end)
+    end
+
+    test "stops on max-failures - concurrently" do
+      in_fixture("max_failures", fn ->
+        assert mix(:concurrent, ~w[test --max-failures 2 --max-cases 1]) =~
+                 "\n3 tests, 2 failures\n"
+
+        assert mix(:concurrent, ~w[test --max-failures 2 --max-cases 2]) =~
+                 "\n4 tests, 2 failures\n"
+
+        assert mix(:concurrent, ~w[test --max-failures 2 --max-cases 3]) =~
+                 "\n5 tests, 2 failures\n"
+
+        assert mix(:concurrent, ~w[test --max-failures 8 --max-cases 10]) =~
+                 "\n12 tests, 3 failures, 5 invalid\n"
+      end)
+    end
+
+    test "aborts concurrent running tests" do
```

Comment:
This test and the other two below are excellent. Can we please move them to the ExUnit suite? They are really testing ExUnit and not the mix integration. Thanks! :heart: